<a href="https://colab.research.google.com/github/kiminlim/ds-section1-sprint2/blob/master/n121-hypothesis-test/n121a_hypothesis_test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src='https://i.imgur.com/RDAD11M.png' width = '200' align = 'right'>

## *DATA SCIENCE / SECTION 1 / SPRINT 2 / NOTE 1*

# 📝 Assignment

# 가설 검정

## 1. One-Sample T-test 연습 (일원카이제곱 검정)

### 데이터 불러오기
- [서울시 가로수 데이터셋](https://data.seoul.go.kr/dataList/367/S/2/datasetView.do)을 사용합니다.

다음 링크를 참조하세요. 
<https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/seoul_tree/seoul_tree.txt>


In [1]:
# 이 곳에 답안을 작성하시기 바랍니다.

import pandas as pd

trees = pd.read_csv('https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/seoul_tree/seoul_tree.txt',  sep = '\t', skiprows = 1)
trees = trees.replace({'-':0})

print(trees.head())
print(trees.shape) # (28, 24)

     기간  자치구    노선수       수량     은행나무  ...    무궁화    소나무   목련    이팝나무      기타
0  2019  서울시  1,412  307,351  108,050  ...  2,374  4,539  378  20,088  20,112
1  2019  종로구     70    7,329    4,072  ...      0    239    0     711      29
2  2019   중구     48    8,250    3,640  ...     46  2,162    0     470     817
3  2019  용산구     39    7,835    4,444  ...      0     51  126     198     780
4  2019  성동구     38    8,120    1,090  ...      0     53    0     709     646

[5 rows x 24 columns]
(28, 24)


### "서울시에는 구별로 평균 약 400 그루 정도의 `이팝나무`가 있다"라는 가설에 대해 가설검정을 시행하세요. 

- 단, `서울시`, `공원녹지사업소`, `시설관리공단` 부분의 데이터는 제외합니다.
- 가설검정 결과의 pvalue 값을 `pv1`에 저장하세요. (`.pvalue` 사용)

In [2]:
# 이 곳에 답안을 작성하시기 바랍니다

trees['이팝나무'] = pd.to_numeric(trees['이팝나무'].str.replace(',','')) 
trees['이팝나무'].head()

0    20088
1      711
2      470
3      198
4      709
Name: 이팝나무, dtype: int64

In [3]:
trees[ (trees['자치구'] == '서울시') | (trees['자치구'] == '공원녹지사업소') | (trees['자치구'] == '시설관리공단') ].index 

Int64Index([0, 26, 27], dtype='int64')

In [4]:
idx = trees[ (trees['자치구'] == '서울시') | (trees['자치구'] == '공원녹지사업소') | (trees['자치구'] == '시설관리공단') ].index 
trees = trees.drop(idx)
trees.shape

(25, 24)

In [5]:
trees

기간   자치구  노선수      수량    은행나무  양버즘나무  ... 살구나무    무궁화    소나무   목련  이팝나무     기타
1   2019   종로구   70   7,329   4,072  1,071  ...    0      0    239    0   711     29
2   2019    중구   48   8,250   3,640    162  ...   21     46  2,162    0   470    817
3   2019   용산구   39   7,835   4,444  1,422  ...   39      0     51  126   198    780
4   2019   성동구   38   8,120   1,090  2,401  ...    0      0     53    0   709    646
5   2019   광진구   28   7,223   1,247  2,306  ...    0      0     20  113   760    443
6   2019  동대문구   43  10,664   3,749  3,342  ...   26      1      0    0  1176    456
7   2019   중랑구   57   8,365   2,774  2,759  ...    0    212     64    0   449  1,156
8   2019   성북구   78   8,449   2,722  1,282  ...    9      5     97    0   853    547
9   2019   강북구   31   7,789   3,510    934  ...    0      0  1,007    0   575    410
10  2019   도봉구   63   9,572   4,914    323  ...    0    214     15    0    47  1,070
11  2019   노원구   50  14,489   6,505  3,336  ...   58    150     32    1   142  1,135
12  2019   은평구   35   9,880   4,324  1,653  ...    0      3      0    0   443    582
13  2019  서대문구   39   8,627   3,474  1,890  ...    0     70      0    0   493  1,120
14  2019   마포구   92  14,538   3,353  3,581  ...   64      0    124   93  1280    348
15  2019   양천구   69  13,501   5,622  3,992  ...    0    197      0    0   784    595
16  2019   강서구   55  13,719   3,977    706  ...    0     59    282    0  1426    568
17  2019   구로구   81  11,548   3,594  1,527  ...    0      0     30   26  1082  1,210
18  2019   금천구   30   6,603   3,545  1,024  ...    0      0      5    0   184     69
19  2019  영등포구  114  18,220   5,879  3,439  ...    0      0     40   19  1281    651
20  2019   동작구   29   7,083   2,178    829  ...    0    206      7    0   397    794
21  2019   관악구   32  10,154   2,280  2,188  ...    0  1,120     59    0   249  1,194
22  2019   서초구   56  16,548   3,800  5,036  ...    0     91      0    0   446  1,143
23  2019   강남구   60  23,349   7,210  7,559  ...    0      0      0    0   768    913
24  2019   송파구   99  22,858  11,894  4,180  ...    0      0    215    0  1651    450
25  2019   강동구   66  17,152   5,474  4,284  ...    0      0      0    0   642  1,969

[25 rows x 24 columns]

In [6]:
trees['이팝나무'].describe()

count      25.00000
mean      688.64000
std       426.86599
min        47.00000
25%       443.00000
50%       642.00000
75%       853.00000
max      1651.00000
Name: 이팝나무, dtype: float64

In [7]:
# 가설 검정 #*
from scipy import stats
result = stats.ttest_1samp(trees['이팝나무'], 400)   # 데이터, 비교하려는 값

print('t검정 통계량 = %.3f, pvalue = %.3f'%(result)) 

t검정 통계량 = 3.381, pvalue = 0.002


* 귀무가설 : m = 400 ( 이팝나무의 서울시 구별 평균 나무 수는 400그루이다. )
* 대립가설 : m =! 400 ( 이팝나무의 서울시 구별 평균 나무 수가 400그루가 아니다. )

pvalue = 0.002 < 0.05 이므로 유의수준 5%에서 귀무가설 기각, 이는 서울시 구별 이팝나무의 평균 나무 수는 400그루가 아니라는 강한 증거가 된다. )


In [8]:
# t값
( 688.64 - 400 ) / ( 426.866 / 5 )
# 자유도가 24인 t분포에서 검정통계량인 3.38은 0.002의 유의확률을 가진다.

3.3809204762150182

In [9]:
pv1 = result.pvalue
pv1

0.002470712850935819

In [10]:
# trees.iloc[24,:]

---
### 이번에는 10개의 구를 샘플링하여, 위에 대한 가설검정을 다시 한번 시행해보세요. x

- 단, 아래의 코드를 참조하여 sample을 위한 seed 는 123로 설정하세요.
- 가설검정 결과의 pvalue값을 `pv2`에 저장하세요


x -> 샘플링 방법의 문제였음. 
나는 25개의 데이터 중 10개를 뽑기 위해 0~24의 인덱싱 중 10개의 번호를 뽑아 해당 인덱스의 데이터를 가져왔는데 그렇게 안하고 그냥 바로 데이터를 집어넣어도 되었다. 
그런데 그런다고 데이터의 인덱스가 달라지나?

#### 1) 인덱스를 추출해서 샘플링 했을 때

In [11]:
import numpy as np
np.random.seed(123)

 
idx = np.random.choice(24, size = 10, replace=False)  # "~"와 " "을 적절한 값으로 채우시면 됩니다
idx

array([ 5, 20, 21, 18, 15,  8,  7, 11,  4,  3])

In [12]:
trees_10 = trees.iloc[idx,:]

result = stats.ttest_1samp(trees_10['이팝나무'], 400)   # 데이터, 비교하려는 값

print('t검정 통계량 = %.3f, pvalue = %.3f'%(result)) # < 0.05 이므로 귀무가설 기각, 이팝나무의 서울시 자치구별 평균 나무 수가 400그루라는 가설은 유의수준 5%에서  유의하지 않다. 

t검정 통계량 = 3.160, pvalue = 0.012


* 귀무가설 : m = 400 ( 이팝나무의 서울시 구별 평균 나무 수는 400그루이다. )
* 대립가설 : m =! 400 ( 이팝나무의 서울시 구별 평균 나무 수가 400그루가 아니다. )

pvalue < 0.05 이므로 유의수준 5%에서 귀무가설을 기각한다.
( 이는 서울시 구별 이팝나무의 평균 나무 수는 400그루가 아니라는 강한 증거가 된다. )

In [13]:
trees_10.describe()

기간         이팝나무
count    10.0    10.000000
mean   2019.0   791.800000
std       0.0   392.091486
min    2019.0   249.000000
25%    2019.0   478.250000
50%    2019.0   734.500000
75%    2019.0  1095.250000
max    2019.0  1426.000000

In [14]:
import numpy as np

(791.8 - 400) / (392.0915 / np.sqrt(10)) # t값이 3.16과 비슷함을 확인. 자유도가 9인 t분포에서 3.16의 검정통계량은 0.012의 피벨류를 가진다. < 0.05 -> 귀무가설 기각. 

3.1599266682750606

#### 2) 이팝나무 데이터를 바로 샘플링 했을 때

In [15]:

np.random.seed(123)

trees_10 = np.random.choice(trees['이팝나무'], size=10)

In [16]:
trees_10

array([1280,  198,  198,  449,  184,  397,  142,  768,  470,  711])

In [17]:
np.mean(trees_10)

479.7

In [18]:
result = stats.ttest_1samp(trees_10, 400)   # 데이터, 비교하려는 값

print('t검정 통계량 = %.3f, pvalue = %.3f'%(result))

t검정 통계량 = 0.707, pvalue = 0.498


* 귀무가설 : m = 400 ( 이팝나무의 서울시 구별 평균 나무 수는 400그루이다. )
* 대립가설 : m =! 400 ( 이팝나무의 서울시 구별 평균 나무 수가 400그루가 아니다. )

pvalue = 0.498 > 0.05 이므로 유의수준 5%에서 귀무가설을 기각할 수 없다. 즉, 샘플링한 10개의 이팝나무 데이터로는 서울시의 구별 평균 나무 수가 400그루가 아니라는 주장을 뒷받침할 수 없다.

-> 겨우 두번의 다른 방식의 10개의 데이터 표본에서 완전히 반대되는 검정 결과가(p벨류)가 나왔다. 이를 통해 작은 수의 데이터로의 가설 검정은 신뢰할 수 없다는 결론을 도출할 수 있었다.

In [19]:
pv2 = result.pvalue
pv2 #######

0.49772576203804575

In [20]:
print(trees_10.mean(), trees_10.std())

479.7 338.4024379344806


In [21]:
(479.7 - 400) / ( 338.4024379344806 / np.sqrt(10)) 

0.7447745679781923


---
## 2. Two-Sample T-test 연습 (이원카이제곱 검정)

### "서울시의 구별 평균 느티나무수와 왕벚나무의 수는 차이가 없다"라는 가설에 대해 가설검정을 시행하세요.

- 단, `서울시`, `공원녹지사업소`, `시설관리공단` 부분의 데이터는 제외합니다.
- 가설검정 결과의 pvalue를 `pv3`에 저장하세요

In [22]:
trees['느티나무'] = pd.to_numeric(trees['느티나무'].str.replace(',','')) 
trees['왕벚나무'] = pd.to_numeric(trees['왕벚나무'].str.replace(',','')) 


In [23]:
# t = pd.Dataframe( trees['느티나무'], trees['왕벚나무'] )
trees.loc[:, ( ['느티나무', '왕벚나무'] ) ].describe()



느티나무         왕벚나무
count    25.000000    25.000000
mean   1415.120000  1292.000000
std    1036.783862   832.740106
min     185.000000   254.000000
25%     619.000000   602.000000
50%    1057.000000  1096.000000
75%    2231.000000  2095.000000
max    3419.000000  2952.000000

In [24]:
# 복수표본 t분포 
(1415.12-1292 ) / np.sqrt(    (1036.783862**2 / 25) + (832.740106**2 / 25) )
# t값이 같음을 확인.    

0.46292560136489425

In [25]:
# 이 곳에 답안을 작성하시기 바랍니다

result = stats.ttest_ind(trees['느티나무'], trees['왕벚나무'])
print('t검정 통계량 = %.3f, pvalue = %.3f'%(result))
# 귀무가설 채택. 즉, 유의수준 5%에서 서울시 자치구별 느티나무와 왕벚나무의 평균 수는 차이가 유의하게 없다.

t검정 통계량 = 0.463, pvalue = 0.646


* 귀무가설 : m = 0 ( 차이가 없다. = 느티나무와 왕벚나무의 평균이 같다.)

* 대립가설 : m =! 0  (차이가 있다. = 평균 느티나무 수가 왕벚나무 수의 평균은 같지 않다.)

'같지 않다'라는 내용의 대립 가설 이므로 두 평균의 차이가 음수이든 양수이든 상관이 없다. 즉,  양측 가설 검정을 한다.  

 pvalue = 0.646 > 0.05 이므로 유의수준 5%에서 귀무가설을 기각할 수 없다. 즉, 귀무가설 채택한다. 대립가설을 강하게 뒷받침하는 데이터가 아니므로 반대의 경우에도 귀무가설을 기각하지 못한다. #( 즉, 두 나무의 평균 그루 수에 차이가 별로 없다. ) #

In [26]:
pv3 = result.pvalue
pv3 # 0

0.6455096880085703


### Two-Sample One-tailed test(단측검정)를 사용하여 **평균 느티나무수가 왕벚나무 수보다 크다**라는 가설에 대한 가설검정을 시행하세요.
- 해당 가설검정의 결과를 `ans`에 저장하세요.

이 때 `ans`가 가질 수 있는 값은 아래와 같습니다 :
1. "느티나무" ( 느티나무의 수가 통계적으로 유의하게 많은 경우 )
2. "왕벚나무" ( 왕벚나무의 수가 통계적으로 유의하게 많은 경우 )
3. "없음" (통계적으로 유의하지 않은 경우)

다음 링크를 참조하세요.
<https://stackoverflow.com/questions/15984221/how-to-perform-two-sample-one-tailed-t-test-with-numpy-scipy>

> 귀무가설 : m = 0 ( 차이가 없다. = 느티나무와 왕벚나무의 평균에 차이가 없다.)

> 대립가설1 : m > 0  (양수이다  = 평균 느티나무 수가 왕벚나무 수보다 크다.)  
대립가설2 : m < 0  (음수이다  = 평균 느티나무 수가 왕벚나무 수보다 작다.)

(단, 여기서의 m = 느티나무 수 - 왕벚나무 수)

* 함수에 stats.ttest_ind(trees['느티나무'], trees['왕벚나무']) 순서로 입력 시 
- t값이 양수이면 -> (**느티나무 - 왕벚나무**)가 양수라는 의미이고, 이는 느티나무의 수가 더 많다는 뜻이 된다. 
- 반대로 t값이 음수이면 -> (**느티나무-왕벚나무**)가 음수라는 의미이고, 이는 느티나무보다 왕벚나무의 수가 더 많다는 뜻이 된다.   
(단, 피벨류가 유의하게 작게 나와서 귀무가설이 기각되었을 때를 한정해서! 피벨류가 0.05보다 큰데 t값이 음수거나 양수라고 그 차이가 유의하게 더 많다/ 작다고 말할 수 없기 때문)


In [27]:
pv = pv3/2 # 단측검정 pvalue
pv

0.32275484400428517

In [28]:
result[0]

0.4629256014492562

In [29]:
# 레퍼런스코드
if pv < 0.05:
    if np.mean(zelkova) > np.mean(cherry):  # 평균으로 비교
        ans = '느티나무'
    else:
        ans = '왕벚나무'
else:
    ans = '없음'
    
ans

'없음'

In [30]:
# 단측검정일 시에 피벨류는 0.05보다 작지만, 비교하는 두 나무(왕벚>느티 or 왕벚<느티?) 중 어느 나무가 더 큰지를 결정하기 위해선 t통계량을 지표로 삼아야 한다. 

# t통계량으로 비교한 코드
if pv < 0.05:     # p < 0.05 : 귀무가설 기각 됨 -> 대립가설 2개 중 선택해야 함. 
    if result[0] > 0 :           # t값이 양수이면 -> 느티 - 왕벚 > 0 (느티 > 왕벚)
        ans = '느티나무'
    elif result[0] < 0 :         # t값이 음수 이면 -> 느티 - 왕벚 < 0 (느티 < 왕벚)
        ans = '왕벚나무'
else:      # p > 0.05 -> 귀무가설 기각 reject(두 나무 수의 차이가 없다.) 
    ans = '없음'
    
ans

'없음'

## [추가 공부] ttest_ind 함수 이해하기

In [31]:
!pip install --upgrade scipy

Requirement already up-to-date: scipy in /usr/local/lib/python3.7/dist-packages (1.6.3)


In [32]:
!pip install --upgrade scipy
import scipy

Requirement already up-to-date: scipy in /usr/local/lib/python3.7/dist-packages (1.6.3)


In [33]:
!pip install azure-ai-textanalytics --pre

In [34]:
import scipy

from scipy.stats import ttest_ind  

In [35]:
# 이 곳에 답안을 작성하시기 바랍니다


result = stats.ttest_ind(trees['느티나무'], trees['왕벚나무'])

result = result.pvalue / 2
print(' pvalue = %.3f'%(result))

 pvalue = 0.323


* 귀무가설 : m = 0 ( 차이가 없다. = 느티나무와 왕벚나무의 평균에 차이가 없다.)

* 대립가설 : m > 0  (차이가 있다. = 평균 느티나무 수가 왕벚나무 수보다 크다.)

단측 가설 검정 결과 

 pvalue = 0.323 > 0.05 이므로 유의수준 5%에서 귀무가설을 기각할 수 없다. 즉, 귀무가설 채택한다. 대립가설을 강하게 뒷받침하는 데이터가 아니므로 반대의 경우에도 귀무가설을 기각하지 못한다. #( 즉, 두 나무의 평균 그루 수에 차이가 별로 없다. ) #

In [36]:
###-------------------------------위에서 사이파이 업데이트 한 후 '런타임 다시시작' 후 '런타임 모두 실행'해야함!----###

# 함수 내 alternative인자가 greater일때와 less일 때의 p벨류의 합은 1이다. 
stats.ttest_ind(trees['느티나무'], trees['왕벚나무'],alternative ='less').pvalue + stats.ttest_ind(trees['느티나무'], trees['왕벚나무'],alternative = 'greater').pvalue

1.0

In [37]:
# 귀무가설 : 느티나무 = 왕벚나무 (ttest_ind함수는 두 평균값을 비교하는 ttest이므로 귀무가설은 같다로 고정이 되어 있다. )
# 대립가설이 '크다'이니 p값은 분포상 오른쪽을 가르키므로 0.323 / 느티나무의 평균수가 왕벚나무 수보다 크므로 t값은 양수.
result = stats.ttest_ind(trees['느티나무'], trees['왕벚나무'], alternative = 'greater')
print('t검정 통계량 = %.3f, pvalue = %.3f'%(result))

t검정 통계량 = 0.463, pvalue = 0.323


In [38]:
# 대립가설 : 느티나무 < 왕벚나무
result = stats.ttest_ind(trees['느티나무'], trees['왕벚나무'],alternative ='less')
print('t검정 통계량 = %.3f, pvalue = %.3f'%(result))
# 대립가설이 '작다' 이므로 p값은 분포상 왼쪽을 가르키므로 0.677 (= 1-0.323한 >에서의 반대 값) / 느티나무의 평균수가 왕벚나무 수보다 크므로 함수 내에서 평균값이 큰 변수가 먼저 올 경우엔 t값은 양수를 가진다. 

t검정 통계량 = 0.463, pvalue = 0.677


In [39]:
# 대립가설 : 왕벚 < 느티 
result = ttest_ind(trees['왕벚나무'], trees['느티나무'],alternative ='less') # 왕벚 - 느티 < 0 -> 왕벚 < 느티 
print('t검정 통계량 = %.3f, pvalue = %.3f'%(result)) # 분포상 t=-0.463은 0보다 왼쪽에 위치하므로 p값은 0.323 ( 이는 반대의 경우와 대칭인 자유도가 24인 t분포에서 t값이 부호만 다른 0.463일 때와 대칭이므로 같은 p값을 같게 된다. )

t검정 통계량 = -0.463, pvalue = 0.323


In [40]:
# 대립가설 : 왕벚 > 느티?
result = stats.ttest_ind(trees['왕벚나무'], trees['느티나무'],alternative = 'greater')
print('t검정 통계량 = %.3f, pvalue = %.3f'%(result)) # 피벨류만 반대이므로 0.677

t검정 통계량 = -0.463, pvalue = 0.677


## 🔥 **도전과제**
> 3문제 모두 하면 됩니다.



### 1. Compare mean for multiple-sample 
 
다음 링크를 참조하세요. <https://reneshbedre.github.io/blog/anova.html>

구별, 은행나무, 양버즘나무, 느티나무에 대해서 ANOVA Test를 실행하세요.

단, `서울시`, `공원녹지사업소`, `시설관리공단` 부분의 데이터는 제외합니다.

문자열 데이터 전처리

In [42]:
# to numeric
trees['양버즘나무'] = pd.to_numeric(trees['양버즘나무'].str.replace(',','')) 
trees['은행나무'] = pd.to_numeric(trees['은행나무'].str.replace(',','')) 

In [53]:
trees[['은행나무','양버즘나무','느티나무']].describe()
# 일단 그냥 평균만 봐도 
# 은행    4210 
# 양버즘 2449
# 느티    1415
# 로 매우 다르다. 
# 그러나 좀 더 객관적인 판단을 위해 아노바테스트를 진행해보자

은행나무        양버즘나무         느티나무
count     25.000000    25.000000    25.000000
mean    4210.840000  2449.040000  1415.120000
std     2181.951013  1725.205869  1036.783862
min     1090.000000   162.000000   185.000000
25%     3353.000000  1071.000000   619.000000
50%     3749.000000  2188.000000  1057.000000
75%     4914.000000  3439.000000  2231.000000
max    11894.000000  7559.000000  3419.000000

ANOVA test  


*가설 설정*
* 귀무가설 : 은행나무, 양버즘나무, 느티나무의 평균 수는 같다.  
> = **서울시에 있는 은행나무와 양버즘나무, 느티나무의 수가 평균적으로 비슷하다.**    
* 대립가설 : 은행나무, 양버즘나무, 느티나무의 평균 수는 다르다.
> = **서울시에 있는 은행나무와 양버즘나무, 느티나무의 수가 평균적으로 다르다.**

In [47]:
#-- 은행나무, 양버즘나무, 느티나무 ANOVA test --#

fvalue, pvalue = stats.f_oneway(trees['은행나무'], trees['양버즘나무'], trees['느티나무'])
print(fvalue, pvalue) # pvalue가 0에 가까운 값이므로 세 나무의 구별 평균 나무수는 같지 않다. 
print('fvalue = %.3f, pvalue = %.3f'%(fvalue, pvalue))

fvalue = 17.006, pvalue = 0.000


*검정 결과*  
> pvalue = 0.000 < 0.05 이므로 유의수준 5%에서 귀무가설 기각하고, 대안적으로 대립가설 채택한다.  
결과적으로, **서울시에 있는 은행나무와 양버즘나무, 느티나무의 수는 평균적으로 다르다**고 해석할 수 있다.  
애초에 자치구별로 다양한 종류의 나무들의 수를 비슷하게 맞춰서 심는 것이 쉽지 않을 것이라는 생각이 든다.

### 2. Sampling

- 세션 시간에 배웠던 4가지 sample 방식을 구현해보세요.

- 이를 위한 데이터는 1 부터 100 이며, sampling해야 하는 수는 20개 입니다.

- Stratified random sampling을 위한 그룹을 나누는 기준은 1-10, 11-20, ... 과 같은 방식입니다.

- Cluster sampling 을 위한 그룹을 나누는 기준은 5로 나눴을때의 나머지 값입니다. (0,1,2,3,4)

In [44]:
# 이 곳에 답안을 작성하시기 바랍니다

### 3. Data transform
다음 링크를 참조하여  `Inverse transform sampling` 을 구현합니다.

<https://youtu.be/9ixzzPQWuAY>

0 ~ 1 의 값을 동일한 확률로 갖는 `uniform distribution`으로 부터 `binomial distribution`으로 transform 하는 코드를 작성한 뒤, 두가지 결과를 시각화를 통해 비교하세요.


In [45]:
# 이 곳에 답안을 작성하시기 바랍니다